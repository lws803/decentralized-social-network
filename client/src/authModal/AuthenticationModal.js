import React from "react";

import ModalView from "../common/Modal";
import styled from "styled-components";

class AuthenticationModal extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      name: "",
      password: "",
    };
  }

  validation() {
    // TODO: Add validation here
    return true;
  }

  render() {
    return (
      <ModalView {...this.props}>
        <UserForm>
          <div>
            Name:
            <input
              type="text"
              value={this.state.name}
              onChange={event => this.setState({ name: event.target.value })}
            />
          </div>
          <div>
            Password:
            <input
              type="password"
              value={this.state.password}
              onChange={event =>
                this.setState({ password: event.target.value })
              }
            />
          </div>
          <button
            onClick={() => {
              if (this.validation()) this.props.onSignup(this.state);
            }}
          >
            Sign up
          </button>
          <button
            onClick={() => {
              if (this.validation()) this.props.onLogin(this.state);
            }}
          >
            Login
          </button>
        </UserForm>
      </ModalView>
    );
  }
}

export default AuthenticationModal;

const UserForm = styled.div`
  display: flex;
  height: 100%;
  flex-direction: column;
  width: 100%;
  align-items: center;
  // justify-content: space-between;
`;
