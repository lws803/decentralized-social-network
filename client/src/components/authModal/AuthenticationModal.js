import React from "react";
import PropTypes from "prop-types";

import Validator from "jsonschema";
import { Button, Header, Modal } from "semantic-ui-react";

import history from "../../utils/History";
import { AuthSchema } from "../common/Schemas";

class AuthenticationModal extends React.Component {
  constructor(props) {
    super(props);
    this.user = this.props.user;
    this.state = {
      name: "",
      password: "",
      authenticated: this.user.is,
    };
  }

  validation() {
    const v = new Validator.Validator();
    const result = v.validate(this.state, AuthSchema, { propertyName: "user" });
    if (result.errors.length) {
      alert(result.errors.join("\n"));
    }
    return result;
  }

  createUser() {
    if (this.validation().valid)
      this.user.create(this.state.name, this.state.password, ack => {
        if (ack.err) alert(ack.err);
        else {
          this.authUser();
        }
      });
  }

  authUser() {
    if (this.validation().valid)
      this.user.auth(this.state.name, this.state.password, ack => {
        if (ack.err) alert(ack.err);
        else {
          this.props.reload();
        }
      });
  }

  _handleKeyDown(e) {
    if (e.key === "Enter") {
      this.authUser();
    }
  }

  render() {
    return (
      <Modal open={!this.state.authenticated} onClose={() => history.goBack()}>
        <Modal.Header>Login</Modal.Header>
        <Modal.Content>
          <Modal.Description>
            <Header>Username</Header>
            <input
              type="text"
              id="username_field"
              value={this.state.name}
              onChange={event => this.setState({ name: event.target.value })}
              onKeyDown={e => this._handleKeyDown(e)}
            />
            <Header>Password</Header>
            <input
              type="password"
              id="password_field"
              value={this.state.password}
              onChange={event =>
                this.setState({ password: event.target.value })
              }
              onKeyDown={e => this._handleKeyDown(e)}
            />
          </Modal.Description>
        </Modal.Content>
        <Modal.Actions>
          <Button color="black" id="authenticate_user_btn" onClick={() => this.authUser()}>
            Login
          </Button>
          <Button color="black" id="create_new_user_btn" onClick={() => this.createUser()}>
            Sign up
          </Button>
        </Modal.Actions>
      </Modal>
    );
  }
}

AuthenticationModal.propTypes = {
  user: PropTypes.object.isRequired,
  reload: PropTypes.func,
};

export default AuthenticationModal;
