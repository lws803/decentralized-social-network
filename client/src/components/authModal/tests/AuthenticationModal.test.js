import React from "react";
import Enzyme, { shallow, mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { createSerializer } from "enzyme-to-json";

import AuthenticationModal from "../AuthenticationModal";

Enzyme.configure({ adapter: new Adapter() });
expect.addSnapshotSerializer(createSerializer({ mode: "deep" }));

describe("Authentication Modal component", () => {
  const mockUsername = "lws803";
  const mockPassword = "helloworld1234";
  it("renders correctly", () => {
    const user = {
      is: undefined,
      create: () => {},
      auth: () => {},
    };
    const wrapper = shallow(<AuthenticationModal user={user} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("stores information correctly", () => {
    const user = {
      is: undefined,
    };
    const wrapper = mount(<AuthenticationModal user={user} />);
    wrapper
      .find("#username_field")
      .simulate("change", { target: { value: mockUsername } });
    wrapper
      .find("#password_field")
      .simulate("change", { target: { value: mockPassword } });
    expect(wrapper.state("name")).toEqual(mockUsername);
    expect(wrapper.state("password")).toEqual(mockPassword);
  });

  it("authenticates correctly", () => {
    const mockUserAuthCallBack = jest.fn();

    const user = {
      is: undefined,
      auth: mockUserAuthCallBack,
    };
    const wrapper = mount(<AuthenticationModal user={user} />);
    wrapper
      .find("#username_field")
      .simulate("change", { target: { value: mockUsername } });
    wrapper
      .find("#password_field")
      .simulate("change", { target: { value: mockPassword } });

    wrapper.find("#authenticate_user_btn").at(0).simulate("click");
    expect(mockUserAuthCallBack.mock.calls.length).toEqual(1);
  });

  it("creates new users correctly", () => {
    const mockUserCreateCallBack = jest.fn();

    const user = {
      is: undefined,
      create: mockUserCreateCallBack,
    };
    const wrapper = mount(<AuthenticationModal user={user} />);
    wrapper
      .find("#username_field")
      .simulate("change", { target: { value: mockUsername } });
    wrapper
      .find("#password_field")
      .simulate("change", { target: { value: mockPassword } });

    wrapper.find("#create_new_user_btn").at(0).simulate("click");
    expect(mockUserCreateCallBack.mock.calls.length).toEqual(1);
  });
});
