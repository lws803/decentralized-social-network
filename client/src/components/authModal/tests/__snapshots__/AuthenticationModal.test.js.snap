// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authentication Modal component renders correctly 1`] = `
<Modal
  centered={true}
  closeOnDimmerClick={true}
  closeOnDocumentClick={false}
  dimmer={true}
  eventPool="Modal"
  onClose={[Function]}
  open={true}
>
  <ModalHeader>
    Login
  </ModalHeader>
  <ModalContent>
    <ModalDescription>
      <Header>
        Username
      </Header>
      <input
        id="username_field"
        onChange={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <Header>
        Password
      </Header>
      <input
        id="password_field"
        onChange={[Function]}
        onKeyDown={[Function]}
        type="password"
        value=""
      />
    </ModalDescription>
  </ModalContent>
  <ModalActions>
    <Button
      as="button"
      color="black"
      id="authenticate_user_btn"
      onClick={[Function]}
    >
      Login
    </Button>
    <Button
      as="button"
      color="black"
      id="create_new_user_btn"
      onClick={[Function]}
    >
      Sign up
    </Button>
  </ModalActions>
</Modal>
`;
